src/python/pants/source/source_root.py

```diff
@@ -301,33 +254,60 @@ def register_options(cls, register):
             fingerprint=True,
             default=cls._DEFAULT_THIRDPARTY_ROOTS,
             advanced=True,
+            removal_version="1.30.0.dev0",
+            removal_hint="Use --roots instead.",
             help=fixed_help_fmt.format("third-party"),
         )
 
+        register(
+            "--roots",
+            metavar="<list>",
+            type=list,
+            fingerprint=True,
+            # For Python a good default might be the repo root, but that would be a bad default
+            # for other languages, so best to have no default for now, and force users to be
+            # explicit about this when integrating Pants.  It's a fairly trivial thing to do.
+            default=[],
+            advanced=True,
+            help="A list of source root suffixes. A directory with this suffix will be considered "
+            "a potential source root. E.g., `src/python` will match `<buildroot>/src/python`, "
+            "`<buildroot>/project1/src/python` etc. Prepend a `^/` to anchor the match at the "
+            "buildroot.  E.g., `^/src/python` will match `<buildroot>/src/python` but not "
+            "`<buildroot>/project1/src/python`.",
+        )
+
     @memoized_method
     def get_source_roots(self):
         return SourceRoots(self)
 
     def create_trie(self) -> "SourceRootTrie":
         """Create a trie of source root patterns from options."""
-        trie = SourceRootTrie(self.source_root_factory)
+        trie = SourceRootTrie()
         options = self.get_options()
 
-        for category in SourceRootCategories.ALL:
-            # Add patterns.
-            for pattern in options.get("{}_root_patterns".format(category), []):
-                trie.add_pattern(pattern, category)
-            # Add fixed source roots.
-            for path, langs in options.get("{}_roots".format(category), {}).items():
-                trie.add_fixed(path, langs, category)
+        if options.roots:
+            for path in options.roots:
+                trie.add_pattern(path)
+        else:
+            # We need to issue a deprecation warning even if relying on the default values
+            # of the deprecated options.
+            deprecated_conditional(
+                lambda: True,
+                removal_version="1.30.0.dev0",
+                entity_description="the *_root_patterns and *_roots options",
+                hint_message="Explicitly list your source roots with the `roots` option in the "
+                "[source] scope. See https://pants.readme.io/docs/source-roots.",
```

Comment:
It seems like you could output exactly the config snippet they need which might be good if they were realying on implicit defaults in which case they have no clue what their source roots are.

Reply:
Will do, with the caveat that their true source roots might be a subset of those, because of pattern matching.